File: apps/web/src/services/api/room-services.ts
import {
  RoomService,
  AssignServiceInput,
  UpdateAssignedServiceInput,
} from '@nhatroso/shared';
import { apiFetch } from './base';

export const roomServicesApi = {
  listByRoom: async (roomId: string): Promise<RoomService[]> => {
    const res = await apiFetch(`/rooms/${roomId}/services`);
    if (!res.ok) throw new Error('Failed to fetch assigned room services');
    return res.json();
  },

  assign: async (
    roomId: string,
    data: AssignServiceInput,
  ): Promise<RoomService> => {
    const res = await apiFetch(`/rooms/${roomId}/services`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.error?.code || 'Failed to assign service to room');
    }
    return res.json();
  },

  update: async (
    roomId: string,
    id: string,
    data: UpdateAssignedServiceInput,
  ): Promise<RoomService> => {
    const res = await apiFetch(
      `/rooms/${roomId}/services/${id}`,
      {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(data),
      },
    );
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.error?.code || 'Failed to update assigned service');
    }
    return res.json();
  },

  remove: async (roomId: string, id: string): Promise<void> => {
    const res = await apiFetch(
      `/rooms/${roomId}/services/${id}`,
      {
        method: 'DELETE',
      },
    );
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.error?.code || 'Failed to remove assigned service');
    }
  },
};

File: apps/api/server-api/src/controllers/room_services.rs
use loco_rs::prelude::*;
use loco_rs::controller::extractor::auth::JWT;
use axum::{Json, extract::Path};
use uuid::Uuid;

use crate::{
    views::room_services::{AssignServiceParams, UpdateAssignedServiceParams},
    models::room_services::Model as RoomService,
    utils::{auth, error::error_response},
};

pub async fn assign(
    auth: JWT,
    Path(room_id): Path<Uuid>,
    State(ctx): State<AppContext>,
    Json(params): Json<AssignServiceParams>,
) -> Result<Response> {
    let owner_id = auth::get_user_id(&auth)?;
    match RoomService::assign(&ctx.db, owner_id, room_id, params).await? {
        Ok(res) => format::json(res),
        Err((status, code)) => error_response(code, status),
    }
}

pub async fn list_by_room(
    auth: JWT,
    Path(room_id): Path<Uuid>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let owner_id = auth::get_user_id(&auth)?;
    match RoomService::list_by_room(&ctx.db, owner_id, room_id).await? {
        Ok(res) => format::json(res),
        Err((status, code)) => error_response(code, status),
    }
}

pub async fn update(
    auth: JWT,
    Path((_room_id, service_id)): Path<(Uuid, Uuid)>,
    State(ctx): State<AppContext>,
    Json(params): Json<UpdateAssignedServiceParams>,
) -> Result<Response> {
    let owner_id = auth::get_user_id(&auth)?;
    match RoomService::update_assignment(&ctx.db, owner_id, service_id, params).await? {
        Ok(res) => format::json(res),
        Err((status, code)) => error_response(code, status),
    }
}

pub async fn remove(
    auth: JWT,
    Path((_room_id, service_id)): Path<(Uuid, Uuid)>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let owner_id = auth::get_user_id(&auth)?;
    match RoomService::remove_assignment(&ctx.db, owner_id, service_id).await? {
        Ok(deleted) => format::json(serde_json::json!({ "success": true, "deleted": deleted })),
        Err((status, code)) => error_response(code, status),
    }
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("api/v1")
        .add("/landlord/rooms/{room_id}/services", get(list_by_room).post(assign))
        .add("/landlord/rooms/{room_id}/services/{id}", patch(update).delete(remove))
}
